pkg/tbtc/node.go

```diff
@@ -1,17 +1,222 @@
+// VerifySignature verifies if the signature was generated from the provided
+// DKG result has using the provided public key.
+func (drs *dkgResultSigner) VerifySignature(signedResult *dkg.SignedResult) (bool, error) {
+	return drs.chain.Signing().VerifyWithPublicKey(
+		signedResult.ResultHash[:],
+		signedResult.Signature,
+		signedResult.PublicKey,
+	)
+}
+
+// dkgResultSubmitter is responsible for submitting the DKG result to the chain.
+type dkgResultSubmitter struct {
+	chain Chain
+}
+
+func newDkgResultSubmitter(chain Chain) *dkgResultSubmitter {
+	return &dkgResultSubmitter{
+		chain: chain,
+	}
+}
+
+// SubmitResult submits the DKG result along with submitting signatures to the
+// chain. In the process, it checks if the number of signatures is above
+// the required threshold, whether the result was already submitted and waits
+// until the member is eligible for DKG result submission.
+func (drs *dkgResultSubmitter) SubmitResult(
+	result *dkg.Result,
+	signatures map[group.MemberIndex][]byte,
+	startBlockNumber uint64,
+	memberIndex group.MemberIndex,
+) error {
+	config := drs.chain.GetConfig()
+
+	// Chain rejects the result if it has less than 25% safety margin.
+	// If there are not enough signatures to preserve the margin, it does not
+	// make sense to submit the result.
+	signatureThreshold := config.HonestThreshold +
+		(config.GroupSize-config.HonestThreshold)/2
+	if len(signatures) < signatureThreshold {
+		return fmt.Errorf(
+			"could not submit result with [%v] signatures for signature threshold [%v]",
+			len(signatures),
+			signatureThreshold,
+		)
+	}
+
+	onSubmittedResultChan := make(chan uint64)
+
+	subscription := drs.chain.OnDKGResultSubmitted(
+		func(event *DKGResultSubmittedEvent) {
+			onSubmittedResultChan <- event.BlockNumber
+		},
+	)
+
+	returnWithError := func(err error) error {
+		subscription.Unsubscribe()
+		close(onSubmittedResultChan)
+		return err
+	}
+
+	groupPublicKeyBytes, err := result.GroupPublicKeyBytes()
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not extract public key bytes from the result: [%v]",
+				err,
+			),
+		)
+	}
+
+	alreadySubmitted, err := drs.chain.IsDKGResultSubmitted(groupPublicKeyBytes)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not check if the result is already submitted: [%v]",
+				err,
+			),
+		)
+	}
+
+	// Someone who was ahead of us in the queue submitted the result. Giving up.
+	if alreadySubmitted {
+		return returnWithError(nil)
+	}
+
+	// Wait until the current member is eligible to submit the result.
+	eligibleToSubmitWaiter, err := drs.waitForSubmissionEligibility(
+		startBlockNumber,
+		memberIndex,
+	)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf("wait for eligibility failure: [%v]", err),
+		)
+	}
+
+	for {
+		select {
+		case blockNumber := <-eligibleToSubmitWaiter:
+			// Member becomes eligible to submit the result.
+			subscription.Unsubscribe()
+			close(onSubmittedResultChan)
```

Comment:
This is not needed. If we entered here, there is no possibility to enter into the `<-onSubmittedResultChan` at the same time or in the future because we call the chain and exit from the submitter altogether.

Reply:
Are you sure?
If we don't call `subscription.Unsubscribe()` before calling `chainRelay.SubmitDKGResult(...)`, the handler for this particular member will not be removed from `dkgResultSubmissionHandlers` and a goroutine will be started with this handler when we process the result that we are submitting.

And since the handler writes to the channel:
```
		func(event *event.DKGResultSubmission) {
			onSubmittedResultChan <- event.BlockNumber
		},
```
the goroutine will be waiting forever to write to the channel and we will have a goroutine leak.
So in my opinion we should keep the `subscription.Unsubscribe()` before calling `chainRelay.SubmitDKGResult(...)`. An alternative would be to make the channel buffered:
```
onSubmittedResultChan := make(chan uint64, 1)
```

Reply:
Yeah, that makes sense! Let's just write a comment explaining why we need to unsubscribe here.

Reply:
Added a comment in acefbbf3226e423dfd3ce169b9b5b465adf538a8.